pkg/maps/nat/cell.go

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package nat
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/promise"
+	"github.com/cilium/cilium/pkg/tuple"
+	"github.com/cilium/hive/cell"
+)
+
+var MapDisabled = fmt.Errorf("nat map is disabled")
+
+var Cell = cell.Module(
+	"nat-maps",
+	"NAT Maps",
+	cell.Provide(func(lc cell.Lifecycle, cfgPromise promise.Promise[*option.DaemonConfig]) (promise.Promise[NatMap4], promise.Promise[NatMap6]) {
+		var ipv4Nat, ipv6Nat *Map
+		res4, promise4 := promise.New[NatMap4]()
+		res6, promise6 := promise.New[NatMap6]()
+
+		lc.Append(cell.Hook{
+			OnStart: func(hc cell.HookContext) error {
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*60)
+				defer cancel()
+				cfg, err := cfgPromise.Await(ctx)
+				if err != nil {
+					return fmt.Errorf("failed to wait for config promise: %w", err)
+				}
+				if !cfg.EnableNodePort {
+					res4.Reject(fmt.Errorf("nat ipv4: %w", MapDisabled))
+					res6.Reject(fmt.Errorf("nat ipv6: %w", MapDisabled))
+					return nil
+				}
+
+				ipv4Nat, ipv6Nat = GlobalMaps(cfg.EnableIPv4,
+					cfg.EnableIPv6, true)
+				// Maps are still created in daemon.initMaps(...) under the same circumstances
+				// so we just open them here so they can be provided to hive.
+				//
+				// TODO: Refactor ctmap gc Enable() such that it can use the map descriptors from
+				// here so we can move all nat map creation logic into here.
+				if err := ipv4Nat.Open(); err != nil {
+					return fmt.Errorf("open ipv4 nat map: %w", err)
+				}
+				if err := ipv6Nat.Open(); err != nil {
+					return fmt.Errorf("open ipv6 nat map: %w", err)
+				}
+
+				if cfg.EnableIPv4 {
+					res4.Resolve(ipv4Nat)
+				}
+				if cfg.EnableIPv6 {
+					res6.Resolve(ipv6Nat)
+				}
```

Comment:
I am not sure if promises are the ideal way to do optional values. But if we assume they are, then I would also reject the promise in these cases where you are not going to resolve them, that will help the receiving side to not pointlessly wait for a ctx to expire.

Reply:
Fixed this, me and @joamaki also [brought up concerns](https://github.com/cilium/cilium/pull/32152#discussion_r1603131450) with using promises here.  I've created an issue to track this problem more broadly: https://github.com/cilium/cilium/issues/32557.